src/replication/master_control_channel.cpp

```diff
@@ -0,0 +1,144 @@
+  {
+    public:
+      ack_reader_task (cubcomm::channel *chn, cubstream::stream_ack *stream_ack);
+      void execute () override;
+
+    private:
+      std::unique_ptr<cubcomm::channel> m_chn;
+      cubstream::stream_ack *m_stream_ack;
+  };
+
+  ack_reader_task::ack_reader_task (cubcomm::channel *chn, cubstream::stream_ack *stream_ack)
+    : m_chn (std::move (chn))
+    , m_stream_ack (stream_ack)
+  {
+  }
+
+  void ack_reader_task::execute ()
+  {
+    if (!m_chn->is_connection_alive ())
+      {
+	return;
+      }
+    size_t len = sizeof (cubstream::stream_position);
+    cubstream::stream_position ack_sp;
+    css_error_code ec = m_chn->recv ((char *) &ack_sp, len);
+    if (ec != NO_ERRORS)
+      {
+	m_chn->close_connection ();
+	// will get cleared by control_channel_managing_task
+	return;
+      }
+    m_stream_ack->notify_stream_ack (ack_sp);
+  }
+
+  class control_channel_managing_task : public cubthread::task_without_context
+  {
+    public:
+      control_channel_managing_task (master_ctrl &master_ctrl );
+      void execute () override;
+
+    private:
+      master_ctrl &m_master_ctrl;
+  };
+
+  control_channel_managing_task::control_channel_managing_task (master_ctrl &master_ctrl)
+    : m_master_ctrl (master_ctrl)
+  {
+
+  }
+
+  void control_channel_managing_task ::execute ()
+  {
+    m_master_ctrl.check_alive ();
+  }
+
+  master_ctrl::master_ctrl (cubstream::stream_ack *stream_ack)
+    : m_stream_ack (stream_ack)
+  {
+    cubthread::delta_time dt = std::chrono::seconds (10);
+    control_channel_managing_task *ctrl_channels_manager = new control_channel_managing_task (*this);
+    m_managing_daemon = cubthread::get_manager ()->create_daemon_without_entry (dt, ctrl_channels_manager,
+			"control channels manager");
+  }
+
+  master_ctrl::~master_ctrl ()
+  {
+    cubthread::get_manager ()->destroy_daemon (m_managing_daemon);
+
+    for (auto &cr : m_ctrl_channel_readers)
+      {
+	cubthread::get_manager ()->destroy_daemon (cr.first);
+      }
+
+    // we are not the ones resposible for deallocing this
+    m_stream_ack = NULL;
+  }
+
+  void
+  master_ctrl::add (cubcomm::channel &&chn)
+  {
+    std::lock_guard<std::mutex> lg (m_mtx);
+
+    // assure caller's param gets moved from
+    cubcomm::channel *moved_to_chn = new cubcomm::channel (std::move (chn));
+    cubthread::delta_time dt = cubthread::delta_time (0);
+    ack_reader_task *ack_reader = new ack_reader_task (moved_to_chn, m_stream_ack);
+    m_ctrl_channel_readers.push_back (std::make_pair (cubthread::get_manager ()->create_daemon_without_entry (dt,
+				      ack_reader, "control channel reader"), moved_to_chn));
+  }
+
+  void
+  master_ctrl::check_alive ()
+  {
+    std::lock_guard <std::mutex> lg (m_mtx);
+    for (auto it =  m_ctrl_channel_readers.begin (); it !=  m_ctrl_channel_readers.end (); ++it)
+      {
+	if (!it->second->is_connection_alive ())
+	  {
+	    cubthread::get_manager ()->destroy_daemon (it->first);
```

Comment:
what happens with the channel (second)?

Reply:
gets deleted when ack_reader_task destructor is called